File: backend/src/models/Notification.js
/**
 * BooyahX — Notification Model
 *
 * In-app notifications for players.
 * Supports tournament, match, room, result, reward, and system types.
 */

const mongoose = require('mongoose');

const notificationSchema = new mongoose.Schema(
  {
    playerId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Player',
      required: [true, 'Player ID is required'],
      index: true,
    },
    title: {
      type: String,
      required: [true, 'Title is required'],
      trim: true,
      maxlength: [100, 'Title cannot exceed 100 characters'],
    },
    message: {
      type: String,
      required: [true, 'Message is required'],
      trim: true,
      maxlength: [500, 'Message cannot exceed 500 characters'],
    },
    type: {
      type: String,
      enum: {
        values: ['tournament', 'match', 'room', 'result', 'reward', 'system'],
        message: '{VALUE} is not a valid notification type',
      },
      required: [true, 'Type is required'],
    },
    isRead: {
      type: Boolean,
      default: false,
      index: true,
    },
    relatedTournamentId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Tournament',
      default: null,
    },
    relatedMatchId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Match',
      default: null,
    },
  },
  {
    timestamps: true,
  }
);

// Compound index for efficient queries
notificationSchema.index({ playerId: 1, isRead: 1 });
notificationSchema.index({ playerId: 1, createdAt: -1 });

const Notification = mongoose.model('Notification', notificationSchema);

module.exports = Notification;

File: lib/core/models/notification.dart
/// BooyahX — Notification Model
///
/// Represents a single notification entry.
/// Fields are minimal for list views. Will be extended when backend arrives.
library;

/// Types of notifications supported by the app.
enum BooyahXNotificationType {
  tournament,
  match,
  room,
  result,
  reward,
  system,
}

class BooyahXNotification {
  final String id;
  final String title;
  final String message;
  final BooyahXNotificationType type;
  final DateTime timestamp;
  final bool isRead;
  final String? relatedTournamentId;
  final String? relatedMatchId;

  const BooyahXNotification({
    required this.id,
    required this.title,
    required this.message,
    required this.type,
    required this.timestamp,
    this.isRead = false,
    this.relatedTournamentId,
    this.relatedMatchId,
  });

  BooyahXNotification copyWith({
    String? id,
    String? title,
    String? message,
    BooyahXNotificationType? type,
    DateTime? timestamp,
    bool? isRead,
    String? relatedTournamentId,
    String? relatedMatchId,
  }) {
    return BooyahXNotification(
      id: id ?? this.id,
      title: title ?? this.title,
      message: message ?? this.message,
      type: type ?? this.type,
      timestamp: timestamp ?? this.timestamp,
      isRead: isRead ?? this.isRead,
      relatedTournamentId: relatedTournamentId,
      relatedMatchId: relatedMatchId,
    );
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is BooyahXNotification && other.id == id;
  }

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() =>
      'BooyahXNotification(id: $id, title: $title, type: $type, isRead: $isRead)';
}
